### PR TITLE
NF: Patch2Self3

### PR DESCRIPTION
**Pull request by iprajwalreddy**
#### Overview

This PR introduces a new feature, Patch2Self3, an enhanced method for denoising Diffusion MRI using self-supervised learning. Patch2Self3 efficiently reduces the use of primary memory by leveraging memory-mapped files (memmaps), thereby significantly lowering RAM requirements while maintaining high performance. This approach optimizes resource utilization by using small amounts of disk space without impacting performance, offering a more scalable solution compared to the original Patch2Self.

#### Reference Links

This PR has referenced the following research papers:
* https://proceedings.neurips.cc/paper_files/paper/2020/file/bc047286b224b7bfa73d4cb02de1238d-Paper.pdf
* https://openaccess.thecvf.com/content/CVPR2024/papers/Fadnavis_Patch2Self2_Self-supervised_Denoising_on_Coresets_via_Matrix_Sketching_CVPR_2024_paper.pdf

#### Changes Made

- Added Patch2Self3.py file.
- Added test cases for Patch2Self3.
- Modified meson.build file to include Patch2Self3 and tests.

### CODE REVIEW

**Review by skoudoro — COMMENTED** [review]
Hi @iprajwalreddy,

Thank you for this addition!

Before an initial review, I do not think a new module( file) is needed. all this code could be directly in `patch2self.py`. Since both function have the same name, you might have to add a new argument name `version` to the patch2self function and then handle both version inside this function. Also, some functions seem to be duplicated. 

To ease the maintenance, both file(`patch2self` and `patch2self3`) need to be smoothly merge into one file. 

Question: I understand the gain in performance, but will the result be different between the 2 methods. If NO, the current test of patch2self should also work for patch2self3.

Thank you for this and the future update.

**Comment by iprajwalreddy** [comment]
@skoudoro Thank you for your feedback. I'll do the necessary modifications.

File: dipy/denoise/patch2self.py [hunk 1]
```diff
@@ -1,50 +1,73 @@
...
-
-    vol_idx: int
-        The volume number that needs to be held out for training.
+    matrixA_name : str
+        The name of the memmap file containing the matrix A.
+    matrixA_dtype : dtype
+        The dtype of the matrix A.
+    matrixA_shape : tuple
+        The shape of the matrix A.
+    s : int
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:28]: This parameter is not present in the method signature. Please, remove it. Have not gone through all changes, but check for other cases, please.

File: dipy/denoise/patch2self.py [hunk 2]
```diff
@@ -1,50 +1,73 @@
...
+        The number of rows in the sketch matrix.
+    tmp_dir : str
+        The directory to save the temporary files.
+    sketch_rows : int
+        The number of rows in the sketch matrix.
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:33]: Please, respect the order of the parameters in the method signature. Check this in the rest of the methods that have been changed.

**Comment by skoudoro** [comment]
Hi @iprajwalreddy,

What is the status of this PR? is it ready for a deeper review?  Can you address @jhlegarreta comments also?

**Comment by iprajwalreddy** [comment]
@skoudoro Hi Serge! The errors are all fixed. I'm refactoring the code to make it more cleaner and readable. I'll try to push the new code asap.

File: dipy/denoise/patch2self.py [hunk 3]
```diff
@@ -158,6 +149,204 @@ def _extract_3d_patches(arr, patch_radius):
...
+        one of the following: {'ols', 'ridge', 'lasso'}. Otherwise,
+        it can be an object that inherits from
+        `dipy.optimize.SKLearnLinearSolver` or an object with a similar
+        interface from Scikit-Learn:
+        `sklearn.linear_model.LinearRegression`,
+        `sklearn.linear_model.Lasso` or `sklearn.linear_model.Ridge`
+        and other objects that inherit from `sklearn.base.RegressorMixin`.
+        Default: 'ridge'.
+
+    alpha : float, optional
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:175]: There is no default value in the method signature, so this should not be marked optional.

File: dipy/denoise/patch2self.py [hunk 4]
```diff
@@ -158,6 +149,204 @@ def _extract_3d_patches(arr, patch_radius):
...
+        Regularization parameter only for ridge and lasso regression models.
+        default: 1.0.
+
+    version : int, optional
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:179]: Same here, and potentially in other places.

File: dipy/denoise/patch2self.py [hunk 5]
```diff
@@ -158,6 +149,204 @@ def _extract_3d_patches(arr, patch_radius):
...
+        `dipy.optimize.SKLearnLinearSolver` or an object with a similar
+        interface from Scikit-Learn:
+        `sklearn.linear_model.LinearRegression`,
+        `sklearn.linear_model.Lasso` or `sklearn.linear_model.Ridge`
+        and other objects that inherit from `sklearn.base.RegressorMixin`.
+        Default: 'ridge'.
+
+    alpha : float, optional
+        Regularization parameter only for ridge and lasso regression models.
+        default: 1.0.
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:177]: Remove this from the docstring. You can see many commits in the DIPY history that are trying to remove all of these from the code base to remove maintenance burden.

File: dipy/denoise/meson.build [hunk 6]
```diff
@@ -25,7 +25,7 @@ python_sources = ['__init__.py',
   'nlmeans.py',
   'noise_estimate.py',
   'non_local_means.py',
-  'patch2self.py',
+  'patch2self.py'
```
Comment (jhlegarreta) [inline dipy/denoise/meson.build:28]: Why is this removed ? If this is a necessary style change, it should go into a separate style commit.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
Have not reviewer in depth, but there are things that stand out.

Also, concerning all the `Fix` commits in the PR, what do they fix? Unless a meaningful commit message body is added, the change can hardly be assessed, and when reviewing history one would not know why a fix was needed either.

I would suggest reading the DIPY contribution guidelines.

File: dipy/denoise/patch2self.py (hunk 3 above)
Comment (iprajwalreddy) [reply dipy/denoise/patch2self.py:175]: Hello @jhlegarreta. Thank you so much for the review. I'm really sorry about the commit messages, I was putting the commits as just fixes because I was trying to test out the CIs. I will try to keep it as draft next time.
I was also planning on squashing all the commits towards the end. Sry about the confusing commits once again.

I'll also recheck on all the function docstrings once again. Thank you for your inputs.

File: dipy/workflows/denoise.py [hunk 7]
```diff
@@ -96,19 +107,33 @@ def run(
...
+                        bvals,
+                        model=model,
+                        b0_threshold=b0_threshold,
+                        alpha=alpha,
+                        verbose=verbose,
+                        b0_denoising=b0_denoising,
+                        clip_negative_vals=clip_negative_vals,
+                        shift_intensity=shift_intensity,
+                        version=ver,
+                    )
```
Comment (skoudoro) [inline dipy/workflows/denoise.py:136]: Why do you need a `if` here ? you call the exact same function with the same parameters....

File: doc/examples/denoise_patch2self.py [hunk 8]
```diff
@@ -172,5 +176,29 @@
 # References
 # ----------
 #
-# .. footbibliography::
```
Comment (skoudoro) [inline doc/examples/denoise_patch2self.py:175]: too keep

File: doc/examples/denoise_patch2self.py [hunk 9]
```diff
@@ -172,5 +176,29 @@
...
+#
+# .. [Knoll11] F. Knoll, K. Bredies, T. Pock, R. Stollberger, Second order
+#              total generalized variation (TGV) for MRI. Magnetic resonance
+#              in medicine, 65(2), pp.480-491.
+#
+# .. [Fadnavis24] S. Fadnavis, A. Chowdhury, J. Batson, P. Drineas,
+#                    E. Garyfallidis, Patch2Self2: Self-supervised Denoising
+#                    on Coresets via Matrix Sketching, Proceedings of the IEEE/CVF
+#                    Conference on Computer Vision and Pattern Recognition (2024),
+#                    27641-27651.
```
Comment (skoudoro) [inline doc/examples/denoise_patch2self.py:204]: to remove. check how we cite now in other tutorial or on this one actually

**Review by skoudoro — COMMENTED** [review]
Hi @iprajwalreddy,

See below a quick initial review. Thanks

File: dipy/workflows/denoise.py (hunk 7 above)
Comment (iprajwalreddy) [reply dipy/workflows/denoise.py:136]: Thank you for the review @skoudoro. In this case, we need the if condition as patch_radius argument is not supported in patch2self3. It only works in patch2self version 1.

File: dipy/denoise/patch2self.py [hunk 10]
```diff
@@ -235,24 +432,179 @@ def patch2self(
...
+    patch_radius,
+    model,
+    b0_threshold,
+    out_dtype,
+    alpha,
+    verbose,
+    b0_denoising,
+    clip_negative_vals,
+    shift_intensity,
+    version,
```
[the remark this answers is not in the document]
Comment (iprajwalreddy) [reply dipy/denoise/patch2self.py:549]: @skoudoro We need version here as we use it for `_fit_denoising_model` function, where in we get the predicted array from the `_fit_denoising_model` function when version is 1 or else it returns the fitted model if version is 3.
I did not want to again get `cur_x` separately when version is 1. In the case of `version=3`, `cur_x` is done differently when we use it for prediction and hence we need only the fitted model in the `version=3`.

File: dipy/denoise/patch2self.py [hunk 11]
```diff
@@ -235,24 +432,179 @@ def patch2self(
...
+
+    clip_negative_vals : bool
+        Sets negative values after denoising to 0 using `np.clip`.
+
+    shift_intensity : bool
+        Shifts the distribution of intensities per volume to give
+        non-negative values.
+
+    version : int
+        Version 1 or 3 of Patch2Self to use.
```
[the remark this answers is not in the document]
Comment (iprajwalreddy) [reply dipy/denoise/patch2self.py:600]: Please refer to the above message.

File: dipy/denoise/tests/test_patch2self.py [hunk 12]
```diff
@@ -52,6 +52,22 @@ def test_patch2self_random_noise(rng):
...
+    # both clip and shift = True for version 3
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message=msg, category=UserWarning)
+        S0den_clip = p2s.patch2self(
+            S0,
+            bvals,
+            model="ols",
+            clip_negative_vals=True,
+            shift_intensity=True,
+            version=3,
```
[the remark this answers is not in the document]
Comment (iprajwalreddy) [reply dipy/denoise/tests/test_patch2self.py:70]: Would you mind explaining this a bit further, please?

File: doc/references.bib [hunk 13]
```diff
@@ -2065,6 +2065,18 @@ @inproceedings{Fadnavis2020
   url       = {https://proceedings.neurips.cc/paper_files/paper/2020/file/bc047286b224b7bfa73d4cb02de1238d-Paper.pdf},
 }
 
+@inproceedings{Fadnavis2024,
```
Comment (jhlegarreta) [inline doc/references.bib:2068]: Please, read the documentation about where to place references: newer references go closer to top:
https://github.com/dipy/dipy/blob/master/doc/devel/bibliography.rst

Comment (iprajwalreddy) [reply doc/references.bib:2068]: Thank you for pointing it out @jhlegarreta. Will fix it right away.

File: doc/examples/denoise_patch2self.py [hunk 14]
```diff
@@ -152,6 +152,10 @@
 save_nifti("denoised_patch2self.nii.gz", denoised_arr, affine)
 
 ###############################################################################
+# You can also use Patch2Self version 1 to denoise the data by using version
+# argument. The default version is set to 3. To use version 1, you can call
+# Patch2Self as follows:
+#    patch2self(data, bvals, version=1)
```
Comment (jhlegarreta) [inline doc/examples/denoise_patch2self.py:158]: This will not be displayed with an appropriate formatting.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
Only commented on the docstrings.

Check the documentation CI build: the patch set increases the number of warnings:
https://github.com/dipy/dipy/actions/runs/10860661525/job/30141530877#step:5:433

We are trying to remove all warnings.

File: dipy/denoise/patch2self.py [hunk 15]
```diff
@@ -129,33 +119,227 @@ def _extract_3d_patches(arr, patch_radius):
...
+    patch_size = 2 * patch_radius + 1
     dim = arr.shape[-1]
 
-    all_patches = []
+    # Calculate the shape of the output array
+    output_shape = tuple(arr.shape[i] - 2 * patch_radius[i] for i in range(3))
+    total_patches = np.prod(output_shape)
+
+    # Use np.lib.stride_tricks.sliding_window_view for efficient patch extraction
+    from numpy.lib.stride_tricks import sliding_window_view
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py:136]: import at the top of the file. Otherwise, very cool function

File: dipy/denoise/patch2self.py [hunk 16]
```diff
@@ -236,23 +429,190 @@ def patch2self(
...
+            b0_threshold,
+            out_dtype,
+            alpha,
+            verbose,
+            b0_denoising,
+            clip_negative_vals,
+            shift_intensity,
+            version,
+        )
+    else:
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py:452]: else not needed

File: dipy/denoise/patch2self.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (skoudoro) [reply dipy/denoise/patch2self.py:557]: I does not make sense to me. 

you should not have version here.

when you call `_fit_denoising_model` just put `version=1` since there is no other choice in this specific case.

File: dipy/denoise/patch2self.py [hunk 17]
```diff
@@ -324,9 +683,9 @@ def patch2self(
 
     # Insert the separately denoised arrays into the respective empty arrays
     for vol_idx in range(0, data_dwi.shape[3]):
-        denoised_dwi[..., vol_idx] = _vol_denoise(
-            train_dwi, vol_idx, model, data_dwi.shape, alpha=alpha
-        )
+        denoised_dwi[..., vol_idx] = _fit_denoising_model(
+            train_dwi, vol_idx, model, alpha=alpha, version=version
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py:687]: `version=1` instead.   it does not make sense to have other version in `patch2self_version1`

File: dipy/denoise/patch2self.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (skoudoro) [reply dipy/denoise/patch2self.py:608]: refer to my comment above

**Review by skoudoro — COMMENTED** [review]
Hi @iprajwalreddy,

See some below some comment to complete @jhlegarreta review.

After fixing those, it should be ready to go. Thanks!

**Comment by ShreyasFadnavis** [comment]
@iprajwalreddy -- Sorry for the slowness on this! This is really well done! Thanks 👍🏽 

1 thing -- can we keep `sketch_size` as an input parameter for the user? With that name? 
People who will refer to the paper may not see it instantly. Also the default here is set to `30%` if I understand correctly?

That should also be documented more.

After this PR is merged, there is a portion of the algorithm where the QR needs to be run iteratively to determine the sketch size automatically as a 1st step in the algorithm :)

**Comment by iprajwalreddy** [comment]
Thank you @ShreyasFadnavis. Sry for the late reply. Yup, I think I can add the sketch_size as an input also. I'll also try to see if I can document the 30% sketch size.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
This introduces 2 new documentation warnings. Check the documentation build CI and address those warnings, please.

**Comment by ShreyasFadnavis** [comment]
> Thank you @ShreyasFadnavis. Sry for the late reply. Yup, I think I can add the sketch_size as an input also. I'll also try to see if I can document the 30% sketch size.

One more thing, from a design perspective, can we move `count_sketch` to `dipy.stats` and import from there? I believe we also need to add other sketching mechanisms in a follow up PR -- which would make this code will become unnecessarily large. 

Other sketching mechanisms that would be a must:
- Leverage Score-based Sketching (we need this for a noise map)
- Gaussian Sketching (this is especially useful for smaller organs)

Link to these functions: https://github.com/ShreyasFadnavis/patch2self2/tree/main/matrix_sketching

**Comment by iprajwalreddy** [comment]
@jhlegarreta I am still working on the documentation issues. I should hopefully fix it soon. Thank you for letting me know on the warnings.

**Comment by iprajwalreddy** [comment]
@ShreyasFadnavis Maybe we can do a seperate PR for making the sketching mechanisms seperate.

File: dipy/denoise/patch2self.py [hunk 18]
```diff
@@ -145,16 +145,18 @@ def vol_denoise(
     ----------
     data_dict : dict
         Dictionary containing the following
-        data_name : str
+
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:148]: @iprajwalreddy Do not add unnecessary blank lines. Check and remove them.

File: dipy/denoise/patch2self.py [hunk 19]
```diff
@@ -145,16 +145,18 @@ def vol_denoise(
...
+        - data_name : str
```
Comment (jhlegarreta) [inline dipy/denoise/patch2self.py:149]: @iprajwalreddy This was not checked after merging and prior to committing. Same applies to other parameters in this docstring.

File: doc/examples/denoise_patch2self.py (hunk 14 above)
Comment (iprajwalreddy) [reply doc/examples/denoise_patch2self.py:159]: image

It looks okay.

File: dipy/denoise/patch2self.py (hunk 19 above)
Comment (iprajwalreddy) [reply dipy/denoise/patch2self.py:149]: image

I am committing a new code which fixes it like this. Please let me know if you think it requires any changes.

Comment (jhlegarreta) [reply dipy/denoise/patch2self.py]: Check the new commit, as Sphinx may consider the indented block as a note.

File: doc/examples/denoise_patch2self.py (hunk 14 above)
Comment (jhlegarreta) [reply doc/examples/denoise_patch2self.py:159]: My bad then. Sorry.

File: doc/examples/denoise_patch2self.py [hunk 20]
```diff
@@ -68,7 +68,7 @@
 # does not require noise estimation and should work with any kind of diffusion
 # data.
 
-hardi_fname, hardi_bval_fname, hardi_bvec_fname = get_fnames(name="stanford_hardi")
+hardi_fname, hardi_bval_fname, hardi_bvec_fname = get_fnames("stanford_hardi")
```
Comment (skoudoro) [inline doc/examples/denoise_patch2self.py:71]: please, keep  `get_fnames(name="stanford_hardi")`

File: dipy/denoise/tests/test_patch2self.py [hunk 21]
```diff
@@ -76,10 +94,11 @@ def test_patch2self_boundary(rng):
     S0[:10, :10, :10, :10] = 300 + noise[:10, :10, :10, :10]
 
     bvals = np.repeat(100, 20)
-
-    p2s.patch2self(S0, bvals)
-    assert_greater(S0[9, 9, 9, 9], 290)
-    assert_less(S0[10, 10, 10, 10], 110)
+    for version in [1, 3]:
+        extra_args = {"patch_radius": (0, 0, 0)} if version == 1 else {}
```
Comment (skoudoro) [inline dipy/denoise/tests/test_patch2self.py:98]: Please, add a test with `patch_radius` as an int. it seems to fails

**Review by skoudoro — APPROVED** [review]
Thank you for all this update @iprajwalreddy. 

I see a speedup of 2.5x on standford hardi data which is a small data. I suppose the speedup will be higher and bigger data.

it looks good to me and I will go ahead and merge this PR. 

Thanks to all the reviewers

**Comment by skoudoro** [comment]
arf, I click to fast, I wanted to squash....
